analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs

```diff
@@ -0,0 +1,172 @@
+        "IBus",
+        "IMessageBus"
+    ];
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(compilationStartContext =>
+        {
+            if (compilationStartContext.Compilation.ReferencesControllers())
+            {
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    var symbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+                    if (symbol.IsControllerType()
+                        && symbol.GetAttributesWithInherited().Any(x => x.AttributeClass.DerivesFrom(KnownType.Microsoft_AspNetCore_Mvc_ApiControllerAttribute)))
+                    {
+                        CheckApiController(symbolStartContext, symbol);
+                    }
+                }, SymbolKind.NamedType);
+            }
+        });
+
+    private static void CheckApiController(SonarSymbolStartAnalysisContext symbolStartContext, INamedTypeSymbol symbol)
+    {
+        var memberNames = RelevantMemberNames(symbol);
+
+        if (memberNames.Count < 2)
+        {
+            return;
+        }
+
+        var dependencies = new ConcurrentStack<Dependency>();
+
+        symbolStartContext.RegisterCodeBlockStartAction<SyntaxKind>(codeBlockStartContext =>
+        {
+            if (BlockName(codeBlockStartContext.CodeBlock) is { } blockName)
+            {
+                codeBlockStartContext.RegisterNodeAction(c =>
+                {
+                    if (c.Node.GetName() is { } dependencyName && memberNames.Contains(dependencyName))
+                    {
+                        dependencies.Push(new(blockName, dependencyName));
+                    }
+                }, SyntaxKind.IdentifierName);
+            }
+        });
+
+        symbolStartContext.RegisterSymbolEndAction(symbolEndContext =>
+        {
+            var parents = memberNames.ToDictionary(x => x, x => x); // Start with singleton sets
+            foreach (var dependency in dependencies)
+            {
+                Union(parents, dependency.From, dependency.To);
+            }
+
+            var disjointSets = DisjointSets(parents);
+            if (disjointSets.Count > 1)
+            {
+                var secondaryLocations = SecondaryLocations(symbol, disjointSets);
+                foreach (var primaryLocation in LocationIdentifiers<ClassDeclarationSyntax>(symbol))
+                {
+                    var diagnostic = Diagnostic.Create(Rule, primaryLocation, secondaryLocations.ToAdditionalLocations(), secondaryLocations.ToProperties(), disjointSets.Count);
+                    symbolEndContext.ReportIssue(CSharpFacade.Instance.GeneratedCodeRecognizer, diagnostic);
```

Comment:
Educational: Why use the recognizer at this level? Why not the simpler `ReportIssue` overload?

Reply:
No reason, copy-paste thing I guess. I have changed that to the single-parameter overload.